enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,432 @@
+object. Components requiring legacy cloud provider configuration, including
+kubelet and KCM, read the transformed config map.
+
+CCO will **not** be modified. The new functionality will instead be implemented
+as a change to the [Cluster Cloud Controller Manager Operator (CCCMO)][CCCMO],
+functioning in much the same way as CCO's management of the legacy
+configuration.
+
+[CCO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-config-operator_platform-operators-ref
+[transformation-function]: https://github.com/openshift/cluster-config-operator/tree/master/pkg/operator/kube_cloud_config
+[CCCMO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-cloud-controller-manager-operator_platform-operators-ref
+
+### Separate ConfigMaps for legacy and external cloud providers
+
+The cloud configuration is read by multiple components which will not all be
+upgraded simultaneously. If we simply updated the
+`openshift-config-managed/kube-cloud-config` to be compatible with external CCM
+we would risk breaking KCM and kubelets which have not yet been upgraded.
+Therefore we are constrained to write two versions of the managed config map.
+
+CCO will be continue to create `openshift-config-managed/kube-cloud-config`,
+which will remain compatible with the legacy cloud provider. CCCMO will create a
+new `openshift-config-managed/cloud-controller-manager-config` config map, which
+will be compatible with the external CCM. The external CCM will then read its
+configuration from `openshift-cloud-controller-manager/cloud-conf`.
+
+In a future release, we can deprecate and remove CCO's management of
+`openshift-config-managed/kube-cloud-config` and eventually the config map
+itself. We can also remove the `openstack-credentials/openstack-credentials`
+secret, which will no longer be used by anything. We will immediately deprecate
+this in the 4.11 release and remove it in the 4.12 release.
+
+### Preventing upgrade when config cannot be upgraded
+
+If CCCMO is not able to generate `cloud-controller-manager-config` from
+`openshift-config/cloud-provider-config`, it will mark itself as non-upgradable
+with a user-actionable error message explaining how
+`openshift-config/cloud-provider-config` is incompatible.
+
+As we intend to switch to external CCM in 4.11, this change will be implemented
+in OCP 4.10.
+
+### Compatibility constraints on user-managed config
+
+This has implications for the user-managed configuration:
+
+In 4.10, the user-managed configuration **must** be compatible with the legacy
+cloud provider. The user **must not** specify config options which cannot be
+transformed to config for the external cloud provider. If either of these are
+not satisfied we will mark CCCMO as not upgradeable.
+
+In 4.11, the base `cloud-provider-config` **may** include config options which
+are not compatible with the legacy cloud provider. We will have removed the
```

Comment:
This is not exact, we still initialize some parts of the in-tree cloud provider (for volume detach) even when running with external CCM. Currently, the in-tree openstack cloud-provider is quite picky about the options it accepts and will return an error when it sees an option it doesn't know about and prevent KCM to start, meaning that the user managed config **must** remain compatible with the legacy cloud provider.

We're tracking this issue in https://bugzilla.redhat.com/show_bug.cgi?id=2077933, and changing the behavior with https://github.com/kubernetes/kubernetes/pull/109709.

Reply:
Done. I included a link to that PR here.